src/builder.ts

```diff
@@ -0,0 +1,146 @@
+import { z } from "zod";
+import { grab } from "./grab";
+import type { Selection } from "./grab";
+
+type Query = string;
+type Payload<T> = { schema: T; query: Query };
+
+export class BaseResult<T> {
+  query: string;
+  schema: T;
+
+  constructor({ query, schema }: Payload<T>) {
+    this.query = query;
+    this.schema = schema;
+  }
+
+  public value(): Payload<T> {
+    return { schema: this.schema, query: this.query };
+  }
+}
+
+/**
+ * Unknown, comes out of pipe and is starting point for queries.
+ */
+export class UnknownResult extends BaseResult<z.ZodUnknown> {
+  constructor(payload: Payload<z.ZodUnknown>) {
+    super(payload);
+  }
+
+  // filter to an unknown array
+  filter(filterValue = ""): UnknownArrayResult {
+    this.query += `[${filterValue}]`;
+    return new UnknownArrayResult({
+      ...this.value(),
+      schema: z.array(z.unknown()),
+    });
+  }
+
+  grab<
+    S extends Selection,
+    CondSelections extends Record<string, Selection> | undefined
```

Comment:
Just curious - is there a benefit to naming the types up in the generic parameter area versus just putting them in the function argument area below?

Reply:
I've got those in generics because I have conditional types like this:

```ts
type NewType = T extends z.ZodArray<any>
    ? ArrayQuery<AllSelection>
    : EntityQuery<AllSelection>;
```

that depend on them in the function body. The function signature varies as a function of these generic types, which is why I put them in the generics section instead of in the function argument itself.

Reply:
Yeah, the fact that these are generic types is part of what powers the type-inference of the whole thing. If you just specify a type for the function arguments, then you won't get the strict type inference.

It's a bit hard to describe in writing, but if you try moving these from generics into the function arguments, and then open the `builder.test.ts` or `index.test.ts` files and start to view some definitions for various `data` variables, you'll notice you lose out on the inference.

Reply:
Ah, I think I get it. You pass the full type (the thing extending your base types) down to the next function rather than it just resetting to the base types you specify in the function arguments. Lots of mind-melting TS concerns at this level of handing types. 🔥🤔🔥